osu.Game/Rulesets/Edit/Checks/Components/Check.cs

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using osu.Game.Beatmaps;
+
+namespace osu.Game.Rulesets.Edit.Checks.Components
+{
+    public abstract class Check
+    {
+        /// <summary>
+        /// The metadata for this check.
+        /// </summary>
+        public abstract CheckMetadata Metadata { get; }
+
+        /// <summary>
+        /// All possible templates for issues that this check may return.
+        /// </summary>
+        public abstract IEnumerable<IssueTemplate> PossibleTemplates { get; }
+
+        /// <summary>
+        /// Runs this check and returns any issues detected for the provided beatmap.
+        /// </summary>
+        /// <param name="beatmap">The beatmap to run the check on.</param>
+        public abstract IEnumerable<Issue> Run(IBeatmap beatmap);
+
+        protected Check()
+        {
+            foreach (var template in PossibleTemplates)
+                template.Origin = this;
```

Comment:
As discussed, remove this and move the `Check` reference to the `Issue` itself. I'd also call it `Check`, not `Origin`.

Reply:
I've removed this part of the code, but decided to keep `Check` in `IssueTemplate`, as the it's accessible from `Issue` (which I've made a bit more convenient with a property):
```cs
public ICheck Check => Template.Check;
```
This way we don't need to explicitly provide the check for each issue constructed from the template, and instead only provide it to the template `ctor`.